README.md
# tableau_api

[![GitHub version](https://badge.fury.io/gh/eoschools%2Ftableau_api.svg)](https://badge.fury.io/gh/eoschools%2Ftableau_api)
![RSpec](https://github.com/eoschools/tableau_api/actions/workflows/specs.yml/badge.svg)
![RuboCop](https://github.com/eoschools/tableau_api/actions/workflows/rubocop.yml/badge.svg)

Ruby interface to the Tableau API v3.1x.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'tableau_api'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install tableau_api

## Usage

### Basic Authentication
```
client = TableauApi.new(host: 'https://tableau.domain.tld', site_name: 'Default', username: 'ExampleUsername', password: 'ExamplePassword')
client.users.create(username: 'baz')
```

### Trusted Authentication
```
client = TableauApi.new(host: 'https://tableau.domain.tld', site_name: 'Default', username: 'ExampleUsername')
client.auth.trusted_ticket
```

### Workbooks
```
# find a workbook by name
workbook = client.workbooks.list.find do |w|
  w['name'] == 'Example Workbook Name'
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake` to run the tests and linters. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Testing

### Docker

```
docker run -it -d -v $(pwd):/src ruby:2 /bin/bash
docker exec -it CONTAINER_ID /bin/bash -c "cd /src && bundle && rake"
```

### Creating New VCR Cassettes

Cassettes should be self-contained, generated by a single spec file
run in defined order. To make changes to specs, it's best to delete a whole cassette
and rerun the whole spec file, except for auth.yml, since it could be difficult to
generate a trusted ticket from a non-trusted host.

To regenerate all the the cassettes, you'll first need to create the following on the Tableau server:
* *Site*: Default
* *Site*: TestSite
 * *Datasource*: test (this might need to be created from Tableau Desktop)
 * *Username*: test_test

And delete the following if they exist:
* *Group*: testgroup (probably under TestSite)
* *Project*: test_project (probably under Default)
* *Site*: Test Site 2
* *User*: test (probably under TestSite)
* *Workbook*: test
* *Workbook*: testpublish

Set the environment variables below to an administrator account and your Tableau Server hostname.

Then run the commands below:

```
docker run -it -d \
  -v $(pwd):/src \
  -e TABLEAU_HOST -e TABLEAU_ADMIN_USERNAME -e TABLEAU_ADMIN_PASSWORD \
  ruby /bin/bash
docker exec -it CONTAINER_ID /bin/bash -c "cd /src && bundle && rake"
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/civisanalytics/tableau_api. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

tableau_api is released under the [BSD 3-Clause License](LICENSE.txt).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [4.3.2] - 2022-07-27

### Changed/Fixed

- Rework arguments in various places to support ruby 3.0 restrictions.
- Bump version.
- Sic rubocop on Ruby 3 standards
- Fix problem where deleting a non-existent user causes an exception.

### Changed

- Specify major version number of ruby in Docker testing command in README


## [4.2.0] - 2022-02-23

### Changed/Fixed

- Allow update_user to handle full_name and email, and make site_role optional.


## [4.1.0] - 2022-02-17

### Changed/Fixed

- Added Users#remove_user to remove the user from the site.
- Added Users#find_by_username that returns the user with the matching
  username.
- Housekeeping, move from Travis to GA, update gems and ruby.


## [4.0.0] - 2020-11-30

### Changed/Fixed

- Changed interface to Connection#api_get_collection to properly merge a string
  query with the pagination parameters instead of overwriting the pagination
  params. This means you can't pass a Hash `query` parameter to `Jobs#list`
  anymore, but job filtering now works properly because colons in the filter
  will not be URL-encoded.
- Avoid mutating extra argument hash to endpoint methods


## [3.0.0] - 2020-11-11

### Added

- Added Jobs resource


### Changed

- Updated to API version 3.1
  - This is a breaking change for site roles: 
    https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_new_site_roles.htm
- Include full error detail in TableauError message


## [2.0.0] - 2019-08-29

- Updated to API version 2.8, compatible with Tableau Server >= 10.5
- Add `refresh` and `image_preview` methods for workbooks
- Bumped the Ruby versions in the Travis matrix build to 2.3.4, 2.2.7, and

## [1.1.2] - 2017-05-24

### Fixed

- Replaced corrupt `.gem` upload to RubyGems

## [1.1.1] - 2017-05-24 - [YANKED]

### Added

- Added Ruby 2.4.1 to the Travis matrix build

### Changed

- Bumped the Ruby versions in the Travis matrix build to 2.3.4, 2.2.7, and
  2.1.10
- Bumped the Ruby version for development to 2.4.1
- Bumped the RSpec version for development to 3.6
- Bumped the WebMock version for development to 3.0
- Updated the authors

### Fixed

- Updated `TableauApi::VERSION`

## [1.1.0] - 2017-03-02

### Added

- [#2](https://github.com/civisanalytics/tableau_api/pull/2)
  - `TableauApi::Resources::Groups` added to support API calls for
    adding/deleting/updating groups.
  - `TableauApi::Resources::Workbook#remove_permissions` added, including
    support for user and group permissions.
  - `TableauApi::Resources::Workbook#add_permissions` supports group
    permissions.
- [#6](https://github.com/civisanalytics/tableau_api/pull/6)
  Added `Users#update_user`
- [#7](https://github.com/civisanalytics/tableau_api/pull/7)
  Added `Sites#create` and `Sites#delete`

### Changed

- [#3](https://github.com/civisanalytics/tableau_api/pull/3)
  `TableauApi::Resources::Workbook#permissions` now returns existing permissions
  instead of adding new permissions. New permissions can be added with
  `TableauApi::Resources::Workbook#add_permissions`.

### Fixed

- [#4](https://github.com/civisanalytics/tableau_api/pull/4)
  Always parse/return workbook permissions as an array

## [1.0.0] - 2016-06-06

- Initial Release

[Unreleased]: https://github.com/civisanalytics/tableau_api/compare/v1.1.2...HEAD
[1.1.2]: https://github.com/civisanalytics/tableau_api/compare/v1.1.1...v1.1.2
[1.1.1]: https://github.com/civisanalytics/tableau_api/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/civisanalytics/tableau_api/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/civisanalytics/tableau_api/tree/v1.0.0


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'tableau_api'
require 'vcr'

if ENV['TABLEAU_ADMIN_USERNAME'].nil? || ENV['TABLEAU_ADMIN_PASSWORD'].nil?
  puts 'TABLEAU_ADMIN_USERNAME and TABLEAU_ADMIN_PASSWORD must be set to record new VCR cassettes'

  ENV['TABLEAU_ADMIN_USERNAME'] = 'FakeTableauAdminUsername'
  ENV['TABLEAU_ADMIN_PASSWORD'] = 'FakeTableauAdminPassword'
end

ENV['TABLEAU_HOST'] = 'http://localhost:2000' if ENV['TABLEAU_HOST'].nil?

VCR.configure do |config|
  config.cassette_library_dir = 'spec/fixtures/vcr_cassettes'
  config.hook_into :webmock

  config.default_cassette_options = { record: :new_episodes, match_requests_on: %i[path method body query] }

  config.filter_sensitive_data('TABLEAU_ADMIN_USERNAME') { ENV['TABLEAU_ADMIN_USERNAME'] }
  config.filter_sensitive_data('TABLEAU_ADMIN_PASSWORD') { ENV['TABLEAU_ADMIN_PASSWORD'].encode(xml: :text) }
  config.filter_sensitive_data('http://TABLEAU_HOST') { ENV['TABLEAU_HOST'] }

  config.allow_http_connections_when_no_cassette = false

  # only record a whitelist of test site and user elements
  config.before_record do |interaction|
    response = interaction.response
    elements = response.body.scan(/<(?:site|user)\s[^>]+name[^>]+>/)
    sensitive_elements = elements.reject { |e| e.match(/"(Default|TestSite|Test Site 2|test|test_test|TABLEAU_ADMIN_USERNAME)"/) }
    unless sensitive_elements.empty?
      sensitive_elements.each { |e| response.body.gsub! e, '' }
      response.body.gsub!(/totalAvailable="\d+"/, "totalAvailable=\"#{elements.length - sensitive_elements.length}\"")
    end
    raise 'Cassette might contain sensitive data; does a regex need to be updated?' if response.body =~ /civis/i
  end
  config.configure_rspec_metadata!
end

RSpec.configure do |config|
  # Turn deprecation warnings into errors.
  config.raise_errors_for_deprecations!

  # Persist example state. Enables --only-failures:
  # http://rspec.info/blog/2015/06/rspec-3-3-has-been-released/#core-new---only-failures-option
  config.example_status_persistence_file_path = 'tmp/examples.txt'
  config.run_all_when_everything_filtered = true
end

RSpec::Matchers.define :be_a_tableau_id do
  match do |actual|
    actual.match(/\A\w{8}-\w{4}-\w{4}-\w{4}-\w{12}\z/)
  end
end

RSpec::Matchers.define :be_a_token do
  match do |actual|
    actual.match(/\A\w{32}\z/) || actual.match(/\A[\w-]{22}\|\w{32}\z/)
  end
end

RSpec::Matchers.define :be_a_trusted_ticket do
  match do |actual|
    actual.match(/\A[\w-]{24}\z/) || actual.match(/\A[\w\-=]{24}:[\w-]{24}\z/)
  end
end

RSpec.shared_context 'tableau client' do
  let(:client) do
    TableauApi.new(
      host: ENV['TABLEAU_HOST'],
      site_name: 'TestSite',
      username: ENV['TABLEAU_ADMIN_USERNAME'],
      password: ENV['TABLEAU_ADMIN_PASSWORD']
    )
  end
end


spec/resources/users_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe TableauApi::Resources::Users, vcr: { cassette_name: 'users' } do
  include_context 'tableau client'

  describe '#create' do
    # http://onlinehelp.tableau.com/v9.0/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Add_User_to_Site%3FTocPath%3DAPI%2520Reference%7C_____7
    it 'can create a user in a site' do
      user = client.users.create(name: 'test', site_role: 'ExplorerCanPublish')
      expect(user['id']).to be_a_tableau_id
      expect(user).to eq(
        'id' => user['id'],
        'name' => 'test',
        'siteRole' => 'ExplorerCanPublish',
        'externalAuthUserId' => '',
        'authSetting' => 'ServerDefault'
      )
    end

    it 'fails with a bad site_role' do
      expect { client.users.create(name: 'test', site_role: 'foo') }.to raise_error('invalid site_role')
    end
  end

  describe '#list' do
    # http://onlinehelp.tableau.com/v9.0/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Get_Users_on_Site
    it 'can list users in a site' do
      sleep(15) if VCR.current_cassette.recording?
      user = client.users.list.find do |u|
        u['name'] == 'test'
      end
      expect(user['id']).to be_a_tableau_id
      expect(user).to eq('id' => user['id'], 'name' => 'test', 'siteRole' => 'ExplorerCanPublish', 'externalAuthUserId' => '')
    end
  end

  describe '#find_user_by_name' do
    # http://onlinehelp.tableau.com/v9.0/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Get_Users_on_Site
    it 'can find users by name in a site' do
      sleep(15) if VCR.current_cassette.recording?
      user = client.users.find_by_name(name: 'test')
      expect(user['id']).to be_a_tableau_id
      expect(user).to eq('id' => user['id'], 'name' => 'test', 'siteRole' => 'ExplorerCanPublish', 'externalAuthUserId' => '')
    end
  end

  describe '#update_user' do
    # https://onlinehelp.tableau.com/current/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Update_User
    it 'can change the site role of a user in a site' do
      user = client.users.list.find do |u|
        u['name'] == 'test'
      end
      expect(user['siteRole']).to eq('ExplorerCanPublish')
      user_after_change = client.users.update_user(user_id: user['id'], site_role: 'Explorer')
      expect(user_after_change['siteRole']).to eq('Explorer')
    end

    it 'raises an error if the site role is not valid' do
      user = client.users.list.find do |u|
        u['name'] == 'test'
      end
      expect do
        client.users.update_user(user_id: user['id'], site_role: 'foo')
      end.to raise_error(RuntimeError, 'invalid site_role')
    end

    it 'raises an error if the user cannot be found' do
      expect do
        client.users.update_user(user_id: 'foo', site_role: 'Viewer')
      end.to raise_error(RuntimeError, 'failed to get user for update')
    end
  end

  describe '#remove_user' do
    it 'can remove a user from a site' do
      user = client.users.list.find do |u|
        u['name'] == 'test'
      end
      expect(client.users.remove_user(name: user['name'])).to be true
    end
  end
end


spec/resources/groups_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe TableauApi::Resources::Groups, vcr: { cassette_name: 'groups' } do
  include_context 'tableau client'

  let(:test_user_id) do
    user = client.users.list.find do |u|
      u['name'] == 'test'
    end
    user['id']
  end

  describe '#create' do
    # http://onlinehelp.tableau.com/v9.0/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Create_Group
    it 'fails with a bad site_role' do
      expect { client.groups.create(name: 'test', default_site_role: 'foo') }.to raise_error('invalid default_site_role')
    end

    it 'can create a group in a site' do
      group = client.groups.create(name: 'testgroup')
      expect(group['id']).to be_a_tableau_id
      expect(group).to eq('id' => group['id'], 'name' => 'testgroup')
    end
  end

  describe '#list' do
    # http://onlinehelp.tableau.com/v9.0/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Query_Groups
    it 'can list groups in a site' do
      sleep(15) if VCR.current_cassette.recording?
      group = find_group_by_name('testgroup')
      expect(group).to eq('id' => group['id'], 'name' => 'testgroup', 'domain' => { 'name' => 'local' })
      expect(group['id']).to be_a_tableau_id
    end
  end

  describe '#add_user' do
    it 'can add a user to a group' do
      group = find_group_by_name('testgroup')
      expect(client.groups.add_user(group_id: group['id'], user_id: test_user_id)).to be true
    end
  end

  describe '#users' do
    it 'can get the users in a group' do
      group = find_group_by_name('testgroup')
      users = client.groups.users(group_id: group['id'])
      expect(users.map { |u| u['id'] }.include?(test_user_id)).to be true
    end
  end

  describe '#remove_user' do
    it 'can remove a user from a group' do
      group = find_group_by_name('testgroup')
      expect(client.groups.remove_user(group_id: group['id'], user_id: test_user_id)).to be true
    end
  end

  def find_group_by_name(name)
    group = client.groups.list.find do |g|
      g['name'] == name
    end
    group
  end
end


lib/tableau_api/connection.rb
# frozen_string_literal: true

module TableauApi
  class Connection
    API_VERSION = '3.1'

    include HTTParty
    headers 'User-Agent' => "tableau_api/#{::TableauApi::VERSION} Ruby/#{RUBY_VERSION}"

    def initialize(client)
      @client = client
    end

    def post(path, **kwargs)
      self.class.post("#{@client.host}/#{path}", kwargs)
    end

    # if the result is paginated, it will fetch subsequent pages
    # collection can be delimited with a period to do nested hash lookups
    # e.g. objects.object
    def api_get_collection(path, collection, page_number: 1, page_size: 100, **kwargs)
      Enumerator.new do |enum|
        loop do
          query = kwargs.fetch(:query, '')
          query += '&' unless query.empty?
          query += "pageSize=#{page_size}&pageNumber=#{page_number}"
          new_kwargs = kwargs.merge(query: query)
          res = api_get(path, **new_kwargs)

          raise TableauError, res if res.code.to_s != '200'

          # ensure the result is an array because it will not be an array if there is only one element
          [collection.split('.').reduce(res['tsResponse']) { |acc, elem| acc && acc[elem] }].flatten.compact.each do |obj|
            enum.yield obj
          end

          break if res['tsResponse']['pagination'].nil?
          break if page_number >= (res['tsResponse']['pagination']['totalAvailable'].to_i / page_size.to_f).ceil

          page_number += 1
        end
      end
    end

    def api_get(path, **kwargs)
      api_method(:get, path, kwargs)
    end

    def api_post(path, **kwargs)
      new_headers = kwargs.fetch(:headers, {}).merge('Content-Type' => 'application/xml')
      api_method(:post, path, kwargs.merge(headers: new_headers))
    end

    def api_put(path, **kwargs)
      new_headers = kwargs.fetch(:headers, {}).merge('Content-Type' => 'application/xml')
      api_method(:put, path, kwargs.merge(headers: new_headers))
    end

    def api_delete(path, **kwargs)
      api_method(:delete, path, kwargs)
    end

    def api_post_multipart(path, parts, headers)
      headers = auth_headers(headers)
      headers['Content-Type'] = 'multipart/mixed'

      uri = URI.parse(url_for(path))

      req = Net::HTTP::Post::Multipart.new(uri.to_s, parts, headers)

      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(req)
      end
    end

    private

    def api_method(method, path, kwargs)
      # do not attach auth headers or attempt to signin if we're signing in
      new_headers = auth_headers(kwargs.fetch(:headers, {})) unless path == 'auth/signin'
      self.class.public_send(method, url_for(path), kwargs.merge(headers: new_headers))
    end

    def url_for(path)
      "#{@client.host}/api/#{API_VERSION}#{'/' unless path[0] == '/'}#{path}"
    end

    # will attempt to signin if the token hasn't been loaded
    def auth_headers(headers = {})
      h = headers.dup
      h['X-Tableau-Auth'] = @client.auth.token
      h
    end
  end
end


lib/tableau_api/resources/users.rb
# frozen_string_literal: true

module TableauApi
  module Resources
    class Users < Base
      SITE_ROLES = %w[
        Creator
        Explorer
        ExplorerCanPublish
        SiteAdministratorCreator
        SiteAdministratorExplorer
        ServerAdministrator
        Unlicensed
        Viewer
      ].freeze

      def create(name:, site_role: 'Viewer')
        raise 'invalid site_role' unless SITE_ROLES.include? site_role

        request = Builder::XmlMarkup.new.tsRequest do |ts|
          ts.user(name: name, siteRole: site_role)
        end

        res = @client.connection.api_post("sites/#{@client.auth.site_id}/users", body: request)

        res['tsResponse']['user'] if res.code == 201
      end

      def list
        url = "sites/#{@client.auth.site_id}/users"
        @client.connection.api_get_collection(url, 'users.user')
      end

      def find_by_name(name:)
        url = "sites/#{@client.auth.site_id}/users"
        res = @client.connection.api_get_collection(url, 'users.user', **{ query: "filter=name:eq:#{name}" })
        res.first
      end

      def update_user(**parms)
        parms = parms.transform_keys { |k| k.to_s.camelcase(:lower) }
        user_id = parms.delete('userId')
        raise "invalid or missing user_id #{user_id} - #{parms}" if user_id.nil?
        raise 'invalid site_role' unless parms['siteRole'].nil? || SITE_ROLES.include?(parms['siteRole'])

        res = @client.connection.api_get ['sites', @client.auth.site_id, 'users', user_id].join('/')
        raise 'failed to get user for update' if res.code != 200

        user = res['tsResponse']['user']
        # parms["name"] ||= user["name"]
        parms['siteRole'] ||= user['siteRole']
        request = Builder::XmlMarkup.new.tsRequest do |ts|
          ts.user(parms)
        end
        res = @client.connection.api_put(['sites', @client.auth.site_id, 'users', user_id].join('/'), body: request)
        res['tsResponse']['user'] if res.code == 200
      end

      def remove_user(name:)
        user = find_by_name(name: name)
        false if user.nil?

        res = @client.connection.api_delete("sites/#{@client.auth.site_id}/users/#{user['id']}")
        raise 'failed to remove user' if res.code != 204

        res.code == 204
      end
    end
  end
end


lib/tableau_api/resources/workbooks.rb
# frozen_string_literal: true

require 'zip'

module TableauApi
  module Resources
    class Workbooks < Base
      def version(file)
        version = nil

        if File.exist?(file) && File.extname(file) == '.twbx'
          Zip::File.open(file) do |zip_file|
            entry = zip_file.glob('*.twb').first
            version = HTTParty::Parser.new(entry.get_input_stream.read, :xml).parse['workbook']['version']
          end
        end

        version
      end

      # rubocop:disable Metrics/ParameterLists
      def publish(name:, project_id:, file:, overwrite: false, show_tabs: false, connection_username: nil, connection_password: nil, connection_embed: false)
        request = Builder::XmlMarkup.new.tsRequest do |ts|
          ts.workbook(name: name, showTabs: show_tabs) do |wb|
            wb.project(id: project_id)
            wb.connectionCredentials(name: connection_username, password: connection_password, embed: connection_embed) if connection_username
          end
        end

        query = URI.encode_www_form([['overwrite', overwrite]])
        path = "sites/#{@client.auth.site_id}/workbooks?#{query}"

        parts = {
          'request_payload' => request,
          'tableau_workbook' => UploadIO.new(file, 'application/octet-stream')
        }

        headers = {
          parts: {
            'request_payload' => { 'Content-Type' => 'text/xml' },
            'tableau_workbook' => { 'Content-Type' => 'application/octet-string' }
          }
        }

        res = @client.connection.api_post_multipart(path, parts, headers)

        return HTTParty::Parser.new(res.body, :xml).parse['tsResponse']['workbook'] if res.code == '201'

        raise TableauError, res
      end
      # rubocop:enable Metrics/ParameterLists

      CAPABILITIES = %w[
        AddComment ChangeHierarchy ChangePermissions Delete ExportData ExportImage ExportXml
        Filter Read ShareView ViewComments ViewUnderlyingData WebAuthoring Write
      ].freeze

      CAPABILITY_MODES = %w[ALLOW DENY].freeze

      def permissions(workbook_id:)
        res = @client.connection.api_get("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/permissions")

        raise TableauError, res if res.code != 200

        permissions = HTTParty::Parser.new(res.body, :xml).parse['tsResponse']['permissions']['granteeCapabilities']
        return [] if permissions.nil?

        permissions = [permissions] unless permissions.is_a? Array
        permissions.map do |p|
          grantee_type = p['group'].nil? ? 'user' : 'group'

          capabilities = {}
          capabilities_list = p['capabilities']['capability']
          capabilities_list = [capabilities_list] unless capabilities_list.is_a? Array

          capabilities_list.each do |c|
            capabilities[c['name'].to_sym] = c['mode'] == 'Allow'
          end

          {
            grantee_type: grantee_type,
            grantee_id: p[grantee_type]['id'],
            capabilities: capabilities
          }
        end
      end

      # capabilities is a hash of symbol keys to booleans { Read: true, ChangePermissions: false }
      def add_permissions(workbook_id:, user_id: nil, group_id: nil, capabilities:)
        validate_user_group_exclusivity(user_id, group_id)

        request = permissions_request(workbook_id, user_id, group_id, capabilities)
        res = @client.connection.api_put("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/permissions", body: request)

        res.code == 200
      end

      def delete_permissions(workbook_id:, user_id: nil, group_id: nil, capability:, capability_mode:)
        validate_user_group_exclusivity(user_id, group_id)
        raise 'invalid capability' unless CAPABILITIES.include? capability.to_s
        raise 'invalid mode' unless CAPABILITY_MODES.include? capability_mode.to_s

        subpath = user_id ? "users/#{user_id}" : "groups/#{group_id}"
        subpath += "/#{capability}/#{capability_mode.capitalize}"
        res = @client.connection.api_delete("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/permissions/#{subpath}")

        res.code == 204
      end

      def update(workbook_id:, owner_user_id:)
        request = Builder::XmlMarkup.new.tsRequest do |ts|
          ts.workbook(id: workbook_id) do |w|
            w.owner(id: owner_user_id)
          end
        end

        res = @client.connection.api_put("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}", body: request)

        res.code == 200
      end

      def refresh(workbook_id:)
        request = Builder::XmlMarkup.new.tsRequest
        res = @client.connection.api_post("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/refresh", body: request)
        res.code == 202
      end

      def preview_image(workbook_id:)
        res = @client.connection.api_get("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/previewImage")
        res.body if res.code == 200
      end

      def find(workbook_id)
        res = @client.connection.api_get("sites/#{@client.auth.site_id}/workbooks/#{workbook_id}")
        res['tsResponse']['workbook'] if res.code == 200
      end

      def views(workbook_id)
        url = "sites/#{@client.auth.site_id}/workbooks/#{workbook_id}/views"
        @client.connection.api_get_collection(url, 'views.view')
      end

      def list
        url = "sites/#{@client.auth.site_id}/users/#{@client.auth.user_id}/workbooks"
        @client.connection.api_get_collection(url, 'workbooks.workbook')
      end

      private

      def permissions_request(workbook_id, user_id, group_id, capabilities)
        Builder::XmlMarkup.new.tsRequest do |ts|
          ts.permissions do |p|
            p.workbook(id: workbook_id)
            p.granteeCapabilities do |gc|
              gc.user(id: user_id) if user_id
              gc.group(id: group_id) if group_id
              gc.capabilities do |c|
                capabilities.each do |k, v|
                  k = k.to_s
                  raise "invalid capability #{k}" unless CAPABILITIES.include? k

                  c.capability(name: k, mode: v ? 'Allow' : 'Deny')
                end
              end
            end
          end
        end
      end

      def validate_user_group_exclusivity(user_id, group_id)
        raise 'cannot specify user_id and group_id simultaneously' if user_id && group_id
        raise 'must specify user_id or group_id' unless user_id || group_id
      end
    end
  end
end
